### ISSUE TITLE
Make context combination more explicit with `copy()`

### ISSUE THREAD

**Issue by voddan** [Kotlin/kotlinx.coroutines#264]
Proposed changes:
===============
- Add `CoroutineContext.copy(....)`
- Remove the optional `parent` parameter from coroutine builders   
- (Possibly) deprecate `CoroutineContext#plus(CoroutineContext)` to use only in library code

Pros
=====
- very explicit about what it does (e.i. copies the context)
- explicit about the non-commutativity of the operation (`context1 + context2 != context2 + context1)
- explicit about what parts of the context it changes (e.i. `parent` or `name`)

Cons
====
- harder to interact between libraries

Possible implementation
===================

    fun CoroutineContext.copy(
        parent: Job? = null, 
        name: String? = null,
        id: Long? = null,
        from: CoroutineContext? = null
    ) {
        var res = this  // not memory efficient!
        if(from != null) res = res + from  // should go first
        if(parent != null) res = res + parent
        if(name != null) res = res + CoroutineName(name)
        if(id != null) res = res + CoroutineId(id)
        return res
    }

Usage example:
============

Before:
```
launch(context + job) {
   ...
}

launch(context, parent = job) {
   ...
}

launch(parent = job) {
   ...
}

launch(context + anotherContext) {
   ...
}

launch(context + job + anotherContext) {
   ...
}
```

After:
```
launch(context.copy(parent = job)) {
   ...
}

launch(context.copy(from = anotherContext)) {
   ...
}

launch(context.copy(parent = job, from = anotherContext)) {
   ...
}
```

**Comment by jcornaz** [comment]
I think `+` is semantically more correct, because it does assemble elements into a new `CoroutineContext`. But `copy` wouldn't *copy*.
```kotlin
val element1 = CommonPool
val element2 = Job()

val context = element1 + element2 // Here result IS a context with element1 + element2
val copy = element1.copy(parent = element2) // Here copy is not a copy of element1.
```


>explicit about the non-commutativity of the operation (`context1 + context2 != context2 + context1)

The plus operator (`+`) Is usually NOT commutative. Of course it can be (like for numbers) and it is good when it is. But most of the case it is not. For example with strings: `"a" + "b" != "b" + "a"` and with lists: `listOf(1) + listOf(2) != listOf(2) + listOf(1)`

Conceptually, I see `CoroutineContext` like a `Map` of `Element`. And the `+` operator is coherent and consistent with the `+` operator existing on `Map`.

And finally, with `copy` it wouldn't be possible to assemble contexts.
```kotlin
val context1: CoroutineContext = TODO()
val context2: CoroutineContext = TODO()
val result = context1 + context2 // This is not possible with copy
```

**Comment by voddan** [comment]
> Sorry it would be actually possible to see with context.copy(from = anotherContext). So I only find this even more cryptic than the other usages of copy, because: context never contains an other context, from is not a context element and 'from' looks like it indicate the source of the copy but it doesn't

True. I included that parameter because I wanted something like `context.insertAllFrom(anothercontext)`

Frankly, the whole idea of silently overriding context parameters with parameters of another context seems quite strange to me. Can we safely use such an operation on complex contexts at all?

**Comment by jcornaz** [comment]
>Frankly, the whole idea of silently overriding context parameters with parameters of another context seems quite strange to me.

If we see `CoroutineContext` as a `Map`, it think it makes perfectly sense and is natural to understand and to use, because it is the same behavior as the `+` operator on `Map`: `mapOf(Job to job1) + mapOf(Job to job2) == mapOf(Job to job2)`.

I would event like to say that, in my opinion, as long as the `CoroutineContext` is semantically a `Map`, it *should* be possible to combine contexts the same way we combine maps (with the `+` operator).

**Comment by zjuhasz** [comment]
The before and after usage example provided by @voddan is the only place I can actually find a breakdown of what the plus operator does depending on what type of `CoroutineContext` is being added. I do think without a fairly deep understanding of how kotlinx.coroutines works under the hood, the plus operator with contexts can be very confusing. I don't think it needs to be replaced though as it does make sense after reading the relevant source code.

I think two things should be done instead of replacing the plus operator:

1. The plus operator needs better documentation that describes what it does in each of the likely use cases (at least describe what happens when you add a `Job` to a `CoroutineContext`). Currently the only thing I see documented is what happens if you add a `CoroutineDispatcher` to a `CoroutineContext`.
1. Add an optional `dispatcher` parameter that only accepts a `CoroutineDispatcher` (just like the existing `parent` parameter does with `Job`) to `launch()` and `async()` so coroutines can be built in a way that might be more understandable and readable to users of the library that don't have a deep understanding of the internals. And maybe more readable even for people who do understand the internals well in some situations. Example:

```
launch(parent = job, dispatcher = CommonPool) {
   ...
}
```

The documentation should show examples of achieving equivalent results with the plus operator and with the parent and dispatcher parameters, so if people learn what one does they will know what the other does. Perhaps adding optional dispatcher parameter could add confusion because there are more options for building a coroutine but I don't think it is any more confusing than just being able to set the context and the parent either separately or through the plus operator and I don't see why you would have an optional parameter for one but not the other.

**Comment by voddan** [comment]
> ... so coroutines can be built in a way that might be more understandable and readable to users of the library that don't have a deep understanding of the internals

Thank you, @zjuhasz, for demonstrating my point. The current context API is only intuitive to "experienced users with deep understanding of the internals" (a.k.a. kotlinx maintainers). Are we sure that was the goal?

**Comment by jcornaz** [comment]
> 1. The plus operator needs better documentation that describes what it does in each of the likely use cases (at least describe what happens when you add a Job to a CoroutineContext). Currently the only thing I see documented is what happens if you add a CoroutineDispatcher to a CoroutineContext.

As it seems to be source of confusion, the documentation should be indeed improved. I'm sure the maintainers of kotlinx.coroutines would be happy to merge improvement you'd propose to: https://github.com/Kotlin/kotlinx.coroutines/blob/master/coroutines-guide.md#combining-contexts.

> 2. Add an optional dispatcher parameter that only accepts a CoroutineDispatcher (just like the existing parent parameter does with Job) to launch() and async() so coroutines can be built in a way that might be more understandable and readable to users of the library that don't have a deep understanding of the internals. And maybe more readable even for people who do understand the internals well in some situations. Example:

I think you make a very good point and a agree there is no reason, to have `parent` and not `dispatcher` optional argument for coroutine builders. However, I'd rather remove the `parent` argument than adding a new one.

I found very confusing the `parent` argument has been added, and it seems to still confuse people as you might find `launch(context + job)` and `launch(context, parent = job)` in the documentation. Both achieving the same, but not being evident for a newcomer (I think).

And what about `CoroutineExceptionHandler` and `CoroutineName`? should the coroutine builders provide optional arguments for them as well? We'd end up with a lot of optional argument for coroutine builders, and a lot of argument to declare when creating a generic function which start a coroutine.

> The current context API is only intuitive to "experienced users with deep understanding of the internals" (a.k.a. kotlinx maintainers). Are we sure that was the goal?

I don't think it is intuitive only to the maintainers of kotlinx mainteners (I am not one). I think it is intuitive for people who tried to understand the philosophy. Kotlin coroutines are a new "paradigm" in my opinion. Not just a library. And of course it needs to make an effort to grasp the global philosophy behind it. But I have no idea of how work 90% of the internals of kotlinx coroutines, and I don't even know how the contexts works behind the hood. I just understood that context are conceptually `Map` of `ContextElement` and then it immediately makes sense and is natural to combine them with `+`, because we already do that for standard `Map`.

**Comment by zjuhasz** [comment]
> And what about CoroutineExceptionHandler and CoroutineName? should the coroutine builders provide optional arguments for them as well? We'd end up with a lot of optional argument for coroutine builders, and a lot of argument to declare when creating a generic function which start a coroutine.

I was suggesting to add dispatcher because I'm guessing `dispatcher` and `parent` are the two most commonly set aspects of a `CoroutineContext` although I have no significant data to support that hypothesis. I suppose it would still be inconsistent and potentially add confusion. I'm not so sure adding optional parameters for `CoroutineExceptionHandler` and `CoroutineName` would be a bad thing either. There's not necessarily anything wrong with having lots of optional parameters. I don't think those parameters existing in the coroutine builders built into kotlinx.coroutines means you would need to declare them in your own builders. Even if you're making a library on top of kotlinx.coroutines, your library presumably has a more specific purpose and so it would make sense for the coroutine builders in it to have different (potentially fewer) parameters.

With that said I might not mind if `parent` was removed either depending on how the documentation was improved.

**Comment by zjuhasz** [comment]
Now I'm thinking the best option is to do an overload to avoid confusion from redundant options being allowed simultaneously (like being able to change parent by adding to the context and by using the optional parameter).

So one would be:

```
public fun launch(
    context: CoroutineContext = DefaultDispatcher,
    start: CoroutineStart = CoroutineStart.DEFAULT,
    block: suspend CoroutineScope.() -> Unit
): Job
```

The other would be:

```
public fun launch(
    dispatcher: CoroutineDispatcher = DefaultDispatcher,
    parent: Job? = null,
    exceptionHandler: CoroutineExceptionHandler? = null,
    name: CoroutineName? = null,
    start: CoroutineStart = CoroutineStart.DEFAULT,
    block: suspend CoroutineScope.() -> Unit
): Job
```

Having the option to set the parent as its own parameter and simultaneously set it by adding to the context parameter isn't particularly intuitive, but in some situations code can be more readable with appropriately typed and named parameters for each settable part of the `CoroutineContext`. Perhaps overloaded coroutine builders offer the best of both worlds.

**Comment by jcornaz** [comment]
I'd indeed prefer to have overloads of `launch` than having a `copy` method of on `CoroutineContext`.

But I don't like the idea of increasing the arity of theses functions up to 6. This is basically increasing the coupling of `launch` and `async` to the new types they have to take as argument. It also increase the number of responsibility of theses functions (including the responsibility to build a context, which should not be part of the `launch`and `async` function IMO).

And it is also by design a list of argument which is meant to grow as new context elements are added.

Long story short: Long argument list is a [code smell](https://en.wikipedia.org/wiki/Code_smell)

**Comment by zjuhasz** [comment]
> Long story short: Long argument list is a code smell

In Java yes, but I do not think that applies to languages with named optional parameters like Kotlin as [this blog post](https://blog.kotlin-academy.com/effective-java-in-kotlin-item-2-consider-a-builder-when-faced-with-many-constructor-parameters-1927e69608e1) expresses in detail. Keep in mind this isn't actually changing the functionality of the method, it is just offering a different syntax for building a context.

**Comment by jcornaz** [comment]
>In Java yes, but I do not think that applies to languages with named optional parameters like Kotlin

I aggree it is less bad in languages with named optional parameters, as the usage of the function remains easy an clear. But the biggest part of the smell is that the function as too much responsibilities and too much execution paths that have to be tested. And this apply even with named  named optional parameters.

> Keep in mind this isn't actually changing the functionality of the method, it is just offering a different syntax for building a context.

It does change the functionality of the method. Now the method has to build a context. This was not his responsibility before.

**Comment by zjuhasz** [comment]
> It does change the functionality of the method. Now the method has to build a context. This was not its responsibility before.

OK that's true. It's actually more adding another method which would just build a context and then call the coroutine builder which takes a context. But it only has 1 responsibility, it is a context builder. I don't think builders are a bad practice.

**Comment by jcornaz** [comment]
>I don't think builders are a bad practice.

Me neither. The only think is we already have a context builder. Building context is the very purpose of the `+` operator.

However I'd have nothing against a more explicit builder like this:
```kotlin
fun CoroutineContex(
    dispatcher: CoroutineDispatcher = DefaultDispatcher,
    parent: Job? = null,
    exceptionHandler: CoroutineExceptionHandler? = null,
    name: CoroutineName? = null
): CoroutineContext {
  var context= dispatcher
  
  parent?.let { context += it  }
  exceptionHandler?.let { context += it }
  name?.let { context += it }

  return context
}
```

I'd still prefer to use `+` personally, but I imagine, such builder could be useful fore more explicit context creation.

**Comment by zjuhasz** [comment]
I'd be fine with this solution, it would be slightly easier to maintain in the library but would also look a bit less pretty at the call site. I think the choice is fairly subjective compared to an overload.

**Comment by jcornaz** [comment]
> would also look a bit less pretty at the call site

Yes that's why I'll still use `+` myself. I find this quite pretty:
`launch(dispatcher + job + name + exceptionHandler) { ... }`

But I aggree this is fairly subjective.

**Comment by voddan** [comment]
>>would also look a bit less pretty at the call site
>
>Yes that's why I'll still use + myself. I find this quite pretty:
>launch(dispatcher + job + name + exceptionHandler) { ... }

In my experience I've always found it beneficial to extract coroutineContext into a (class) property for possible reuse and thorough documentation. So the code becomes:

    val wellNamedContext = CoroutineContext(name = "", exceptionHandler = "", parent = job)
    ...
    launch(wellNamedContext) { ... }
 
That eliminates the requirement to have a concise syntax for context creation and allows us to focus on readability rather than brevity.

What do you guys use - inline context composition or constructing the context somewhere else?

**Comment by voddan** [comment]
P.S. 
The only exception to my post above is (presumable) `parent = job`, because it is supposed to change from coroutine to coroutine:

    launch(A) {
        launch(A + coroutineContext) {
        ....
        }
    }

AFAIK that was the reason the `parent` was introduced in the first place.

**Comment by jcornaz** [comment]
If I have to reuse context, I also put it in a property. But I build the context as usual with the `+` operator.

That being said, it is pretty rare. Most of the time it is easier, and more expressive to do it when the coroutine is started (that way the reader doesn't have to guess what's in the context).

I don't think readability is incompatible with brevity. I know it is subjective, but I find the following fairly easy to read:
```kotlin
val wellNamedContext = dispatcher + exceptionHandler + job
```

**Comment by qwwdfsad** [comment]
Proposed change makes context combinations less readable, does not add any value, and `copy` name is kind of misleading because it does not copy anything.
This discussion could have a place in the corresponding KEEP, but now it is too late anyway.

And commutativity is not a property of `plus` operator in general.
